org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/archetype/ArchetypePlugin.java

```diff
@@ -74,8 +77,8 @@
  *
  * @author Eugene Kuleshov
  */
-@Component(service = ArchetypePlugin.class)
-public class ArchetypePlugin {
+@Component(service = {ArchetypeManager.class, ArchetypePlugin.class})
```

Comment:
Not possible, as the ArchetypePlugin is the internal API which is used. If you would only register as ArchetypePlugin you would need to cast in M2EUIPluginActivator. How do you want to get hold of this without casting?

Reply:
> DS allows to replace one service implementation with another one!
This has nothing to do with DS, anyways you would be most likely wrecked then anyways. If you want be extra you can assert the type with instanceof and throw an exception otherwise...

> Also remember that you previously registered only to the internal API ArchetypePlugin via DS, so that pattern is not something I introduced.

Many thing in m2e need improvements and some are just for migrating thing so we should not take this as good examples when refactoring/rework things.

Reply:
In Softwaredevelopment "later" often means "never" :-)

I would prefer to have a cast on the cases where it is required (that is we use no API method) as this is already done elsewhere in m2e and shows where we have an API gap...